syntax = "proto3";

package vs.manas.dt;

message DataBase {
  InstructionSet isa = 1;
  InstructionPackage pkg = 2;
}

message InstructionSet {
  string platform = 1;
  int32 instr_bitwidth = 2;
  int32 instr_code_bitwidth = 3;
  repeated InstrctionTemplate instruction_templates = 4;
}

message InstrctionTemplate {
  string name = 1;
  int32 code = 2;
  repeated string machines = 3;
  repeated SegmentTemplate segment_templates = 4;
}

message SegmentTemplate {
  string name = 1;
  int32 bitwidth = 2;
  bool is_signed = 3;
  int32 default_val = 4;
  repeated IntStrMap verbo_map = 5;
  string comment = 6;
}

message InstructionPackage {
  string name = 1;
  string version = 2;
  string platform = 3;
  repeated InstructionList instruction_lists = 4;
}

message InstructionList {
  string label = 1;
  repeated Instruction instructions = 2;
}

message Instruction {
  string name = 1;
  repeated StrIntMap value_map = 2;
  repeated StrIntMap int_info_map = 3;
  repeated StrStrMap str_info_map = 4;
}

message StrIntMap {
  string key = 1;
  int32 val = 2;
}

message StrStrMap {
  string key = 1;
  string val = 2;
}

message IntStrMap {
  int32 key = 1;
  string val = 2;
}

message ASMProg {
  repeated string contents = 1;
  map<string, ASMRecord> records = 2;
}

message ASMRecord {
  enum Kind {
    UNKNOWN = 0;
    CELL = 1;
    INSTR = 2;
  }
  string id = 1;
  Kind kind = 2;
  string name = 3;
  map<string, string> parameters = 4;
}
